### PR TITLE
Fix @ckeditor Observable, Emitter and EventInfo interfaces 

### PR DESCRIPTION
**Pull request by fedemp**
Please notice that approximately only half the files are actually modified, cause one half belong to the previous version of each package.

* `Observer` implements `Emitter`, so whatever class implements `Observer`, must also implement `Emitter`. https://github.com/ckeditor/ckeditor5/blob/860ac48a77c38ce75402bef91e8ff9cf278aeb40/packages/ckeditor5-utils/src/observablemixin.js#L709
* Classes that used to read `implements Emitter, Observable` now only read `implements Observable` given the previous point.
* Classes that extend `ContextPlugin`, `Command` and `Plugin` do not implement `Observable` and `Emitter` cause their parent class takes care of that. https://github.com/ckeditor/ckeditor5/blob/860ac48a77c38ce75402bef91e8ff9cf278aeb40/packages/ckeditor5-core/src/contextplugin.js#L28, https://github.com/ckeditor/ckeditor5/blob/860ac48a77c38ce75402bef91e8ff9cf278aeb40/packages/ckeditor5-core/src/command.js#L25, and https://github.com/ckeditor/ckeditor5/blob/860ac48a77c38ce75402bef91e8ff9cf278aeb40/packages/ckeditor5-core/src/plugin.js#L17
* `EditorWithUI` is the same as `Editor` but with the `ui` property. https://github.com/ckeditor/ckeditor5/blob/860ac48a77c38ce75402bef91e8ff9cf278aeb40/packages/ckeditor5-core/src/editor/editorwithui.jsdoc#L10-L22
* Generics for `EventInfo` have been improved so now the first optional parameter is a string that is used for the `name`, and the second optional parameter is the class that extended `Emitter`.
* The actual data for `on()` and `listenTo()` has been marked as `unknown`, cause it depends on what event is being listened and what arguments are passed to `fire()`. 

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by fedemp** [comment]
@bendemboski You may want to look at this, cause the generics for `EventInfo` are a breaking change.

File: types/ckeditor__ckeditor5-engine/src/conversion/upcastdispatcher.d.ts [hunk 1]
```diff
@@ -18,10 +17,10 @@ export interface UpcastConversionData {
 }
 
 export interface UpcastConversionApi {
-    consumable: ViewConsumable;
-    schema: Schema;
-    store: Record<string, any>;
-    writer: Writer;
+    consumable?: ViewConsumable;
```
Comment (bendemboski) [inline types/ckeditor__ckeditor5-engine/src/conversion/upcastdispatcher.d.ts:20]: Making these four optional introduced a whole bunch of errors into my plugins' upcasters. I'm sure that they are guaranteed to be set in the places where I'm using them (for example, the last example [here](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_conversion_upcasthelpers-UpcastHelpers.html#function-elementToElement) -- when can they be undefined? Do we need a separate interface or something for the cases where they can be undefined? I don't think we want to force everyone to do null checks in all their upcasters that use the conversion api.

**Review by bendemboski — CHANGES_REQUESTED** [review]

**Comment by bendemboski** [comment]
@fedemp I left a review with a comment about the upcast conversion API, but aside from that this worked well! I pointed my app at it and the observable/emitter/event info changes look great! They allowed me to remove all of my `@ts-expect-error`s related to `listenTo()`, `on()`, and the like.

**Comment by fdintino** [comment]
This isn't a comment on this PR specifically, but I wasn't sure where else to pose my question. I've had to fork the types for `@ckeditor/ckeditor5-engine` for a project I'm working on. One improvement I made, for instance, was to use  template literal types for the downcast and upcast dispatcher event methods, so that the callback arguments get the correct type inferences without needing to cast anything. Another improvement I made was to setup methods like `Node.is` with type predicate signatures, so that you can do idiomatic checks like `if (item.is("element"))` and have the correct type inference in the conditional body. You can see my work [here](https://github.com/fdintino/DefinitelyTyped/tree/ckeditor5-engine-fixes) (fdintino/DefinitelyTyped@5536bf9).

I've noticed though that these type packages are under pretty active development, and there are several open pull requests that will create conflicts once merged. I wanted to reach out to see whether there's interest in incorporating my changes, and if so how best to go about contributing them.

File: types/ckeditor__ckeditor5-engine/src/conversion/upcastdispatcher.d.ts (hunk 1 above)
Comment (fedemp) [reply types/ckeditor__ckeditor5-engine/src/conversion/upcastdispatcher.d.ts]: I'm sure I made them optional based on upstream tests, that sometimes take some liberties.

I've reverted this change.

**Comment by fedemp** [comment]
@fdintino Wow. Your modifications are impressive. I specially liked all the work you put on the `is()` method. Please create a pull request so @bendemboski and me can review it.

**Comment by fedemp** [comment]
@fdintino It seems your modifications will mostly conflict with my modifications to `on()` and related. Do you want to fork this PR and fix those, and then create another PR for your improvements for `is()`?

**Comment by fdintino** [comment]
@fedemp Perhaps it would be better for me to first open a PR for the `is()` improvements. By the time I've written tests for it, this PR will probably be merged, at which point I can move on to resolving any conflicts on the remaining changes for a second PR.

File: types/ckeditor__ckeditor5-cloud-services/src/token/token.d.ts [hunk 2]
```diff
@@ -1,11 +1,14 @@
...
 import { BindChain, Observable } from "@ckeditor/ckeditor5-utils/src/observablemixin";
+import { PriorityString } from "@ckeditor/ckeditor5-utils/src/priorities";
 
 export default class Token implements Observable {
     value: string;
 
     constructor(
         tokenUrlOrRefreshToken: string | (() => string),
-        options: { initValue?: string | undefined; autoRefresh?: boolean | undefined },
+        options?: { initValue?: string | undefined; autoRefresh?: boolean | undefined },
```
Comment (fedemp) [inline types/ckeditor__ckeditor5-cloud-services/src/token/token.d.ts:11]: https://github.com/ckeditor/ckeditor5/blob/d884b5c1b9e7514ba4805e110e5f06106adff7ff/packages/ckeditor5-cloud-services/src/token/token.js#L34

**Comment by fdintino** [comment]
It just occurred to me: would it be better to use `any` instead of `unknown`? With `unknown`, it requires that developers use `as` casting, whereas with `any` developers could simply use type annotations in the callback signatures.

**Comment by fdintino** [comment]
It would also be a breaking change for the places where you previously had `any`, if developers were using type annotations on callback arguments. Since any can be converted to any other type, this would compile, but with unknown they would need to use unknown in the callback signature and then cast with `as` in the callback body.

**Comment by fedemp** [comment]
The "problem" with `any` is that not only we lose type checking, so we also encourage developers to not pay attention to typings.

Even though developers are forced to use casting, the would win on two sides. First, they would win typings back so an any doesn't provide autocomplete. Second, it makes it clear what type you are actually working on.

**Comment by fdintino** [comment]
Another thing that occurs to me (sorry I didn't put all these comments together) is that using `unknown` in the mixins limits our ability to give explicit types in classes that implement Emitter. And, generally speaking, classes that implement Emitter in the CKEditor API have consistently typed data for callbacks, and so could be typed as such here. To use a random example `FileUploader` in `@ckeditor/ckeditor5-cloud-services` `src/uploadgateway/fileuploader.js` should be typed:

```typescript
export class FileUploader {
    on<N extends keyof EventCallbackTypes>(
        event: N,
        callback: (info: EventInfo<N>, ...data: EventCallbackTypes[N]) => void,
        options?: { priority?: number | PriorityString | undefined },
    ): void;
    on<N extends string>(
        event: N,
        callback: (info: EventInfo<N>, ...data: any[]) => void,
        options?: { priority?: number | PriorityString | undefined },
    ): void;
}
```

Using `unknown` in `Emitter` would prevent improving the types in this way (and, more top of mind for me, would prevent me from making the planned improvements to the upcast and downcast dispatchers.

**Comment by fdintino** [comment]
btw I would be happy to do the work of adding these more explicit event listener types across the board.

**Comment by fedemp** [comment]
> btw I would be happy to do the work of adding these more explicit event listener types across the board.

That would be actually great, specially given even though I'm creating these types, I'm not using ckeditor in production yet.

**Comment by fdintino** [comment]
What are your thoughts on changing the `unknown[]` to `any[]` in this pull request, to lay the groundwork for typing event handlers? Otherwise I'll have to remove the `implements Emitter` from classes as I go through and add those types.

**Comment by fedemp** [comment]
> What are your thoughts on changing the `unknown[]` to `any[]` in this pull request, to lay the groundwork for typing event handlers? Otherwise I'll have to remove the `implements Emitter` from classes as I go through and add those types.

If that would help in your improvements, sure! I'll do a search and replace now.

**Review by fdintino — APPROVED** [review]

File: types/ckeditor__ckeditor5-utils/src/dom/emittermixin.d.ts [hunk 3]
```diff
@@ -1,32 +1,66 @@
...
 import EventInfo from "../eventinfo";
 import { PriorityString } from "../priorities";
 
 export interface Emitter extends BaseEmitter {
-    delegate(...events: string[]): EmitterMixinDelegateChain;
-    fire(eventOrInfo: string | EventInfo, ...args: any[]): any;
-    listenTo(
-        emitter: Emitter,
-        event: string,
-        callback: (info: EventInfo, data: DomEventData) => void,
```
Comment (fdintino) [inline types/ckeditor__ckeditor5-utils/src/dom/emittermixin.d.ts:12]: This callback signature was incorrectly copied to all `listenTo` and `on` methods, but here it was actually correct.

**Comment by fdintino** [comment]
It looks like there are emittermixin methods that need to be added to Observable. I pointed my repo to the types on this branch and am getting `Property 'listenTo' does not exist on type 'ImageLoadObserver'` on a line that looks like this:

https://github.com/ckeditor/ckeditor5/blob/daffcd78d94d3991ed4a99d746d702dff45e1365/packages/ckeditor5-image/src/image/imageloadobserver.js#L27

File: types/ckeditor__ckeditor5-utils/src/dom/emittermixin.d.ts (hunk 3 above)
Comment (fedemp) [reply types/ckeditor__ckeditor5-utils/src/dom/emittermixin.d.ts:12]: I've reverted this.

**Comment by fedemp** [comment]
> It looks like there are emittermixin methods that need to be added to Observable. I pointed my repo to the types on this branch and am getting `Property 'listenTo' does not exist on type 'ImageLoadObserver'` on a line that looks like this:
> 
> https://github.com/ckeditor/ckeditor5/blob/daffcd78d94d3991ed4a99d746d702dff45e1365/packages/ckeditor5-image/src/image/imageloadobserver.js#L27

Good catch. `DomEmitterMixin` is mixed into this class but that is not mentioned in the JSDoc. I've updated that too.

**Pull request closed without merging**